src/Chapter2/Exam.hs

```diff
@@ -0,0 +1,265 @@
+module Chapter2.Practice3 where
+import Data.List
+import Data.Char
+
+
+score :: [(String,[(String,Int)])]
+score=[ ("Mojo", [("English", 84), ("Chemisty", 80), ("Physics", 95), ("Geography", 75)])
+    , ("John Doe", [("Chemisty", 80), ("Physics", 95), ("Geography", 75)])
+    , ("Captain Jack", [("Chemisty", 66), ("Phsyics", 33), ("Geography", 56)]) -- Note the spelling error in "Physics"
+    , ("John Doe", [("Chemisty", 90), ("Economics", 45), ("Geography", 56)]) -- Note that "John Doe" is a repeat entry
+    , ("Bahubali", [("Hindi", 45), ("Biology", -90), ("Geography", -75)]) -- Note the negative marks in "Biology" & "Geography"
+    , ("Dexter", [("Tamil", 110), ("Biology", 100), ("Geography", 100)]) -- Note that marks in "Tamil" are greater than 100
+    ]
+
+validsubject :: [String]
+validsubject = ["English", "Geography", "Physics", "Chemistry", "Economics", "Computer Science"]
+
+type InputScore = [(String,[(String,Int)])]
+type Subject = String
+type SubjectAverage = (String,Float)     
+
+averageMarks :: InputScore ->ValidSubjects -> Subject -> SubjectAverage             --1st
+averageMarks score validsubject subject =let listSubMarks =  (filter (\x -> (fst x == subject)) (foldl' (\arr x -> arr ++ (snd x)) [] (allValidNames score validsubject) ) )
```

Comment:
- use pattern matching in lambdas
- don't use a `foldl'` where another higher-order function would be more appropriate and would communicate intent more clearly. In this case, you can replace `foldl'` with `map`. Similarly, use `sum` instead of `foldl'` below. Understand and use `concatMap` where applicable.
- give interim names to different parts of the computation in a `let..in` binding to make the structure & intent of code clear to the reader, eg.:

```
let onlyScores = ...
     scoresOfGivenSubject = filter _todo onlyScores
     sumOfScores = sum ...
in (fromIntegral sumOfScore) / (fromIntegral (length scoresOfAGivenSubject)
```
```